Encoding of mutable array indexing

dario23 opened the pull request.

dario23 committed:
Add encode_place parameter for whether to encode shared or mutable array indexing

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5467,22 +5476,186 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        let res: vir::Expr = self.cfg_method.add_fresh_local_var(array_elem_ty.clone()).into();
+        let res_val_field = res.clone().field(val_field.clone());
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Mutable)?;
+        let idx_val_int = index.clone().field(vir::Field::new("val_int", vir::Type::Int));
```
`Encoder::encode_value_field` is slightly better than hard-coding `"val_int"`.

Actually, now that I think better at it, it's more serious: this might cause a consistency error if nothing else in Prusti calls `encode_value_field` to generate a `val_int` field. It's rare, but if it happens then the Viper program would not have the`field val_int: Int` declaration that is generated by `Encoder::encode_value_field`.

dario23 replied:
agreed, using [`TyCtxt::mk_ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.mk_ty) now and then encoding the freshly-created `usize` type for the `index field`.

dario23 committed:
Implement mutable array access

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5467,67 +5496,226 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+            vir::PermAmount::Write,
+        );
+
+        stmts.push(vir!{ assert [array_access_pred.clone()] });
```
It should be `exhale`, not `assert`.

dario23 replied:
yes; this is the change mentioned in my lengthy comment to make it work a bit more, will change back of course.

fpoli replied:
I see, now I understand the exhale/assert in the lengthy comment. Yes, it's probably because the magic wand is never applied.

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5467,67 +5496,226 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        // NEEDSWORK: the vir! macro can't do everything we want here..
+        let i_var: vir::Expr = vir_local!{ i: Int }.into();
+
+        let zero_le_i = vir!{ [ vir::Expr::from(0) ] < [ i_var.clone() ] };
```
The case `i == 0` should be allowed.

dario23 replied:
good catch :-)

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5467,67 +5496,226 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        stmts.push(vir::Stmt::Assert(
+            vir::Expr::predicate_access_predicate(array_pred, encoded_base_expr, vir::PermAmount::Read),
+            vir::Position::default(),
+        ));
```
Why is this assertion generated? It seems redundant to me.

dario23 replied:
good catch, thanks, will adjust :-)

dario23 committed:
Cleanup

dario23 committed:
Add mutable array referencing testcase

dario23 committed:
Implement in-place array mutation

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4574,6 +4607,89 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+            [old(at.encode_lookup_pure_call(encoded_array.clone(), i_var))]
+        };
+        let forall_body = vir!{ [idx_conditions] ==> [lookup_same_as_old] };
+        let all_others_unchanged = vir!{ forall i: Int :: { /* TODO */ } [ forall_body ] };
```
I'm guessing `lookup_pure(array, i)` should be a good enough trigger here?

dario23 committed:
Add lookup_pure triggers

dario23 committed:
Remove useless `clone()` calls

dario23 committed:
Encode index place in expand_one_level

dario23 committed:
Exhale arrays like local variables in loop permissions

We need to find the maximum used permission for each array, so they are
first collected and then emitted at the end.

dario23 committed:
Add two more testcases

Aurel300 commented on prusti-common/src/vir/borrows.rs:
```diff
@@ -80,11 +80,11 @@ impl Node {
+//     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+//         write!(f, "{:?}", self.borrow)
+//     }
+// }
```
Delete completely or remove the `Debug` derive if you don't need it.

Aurel300 commented on prusti-common/src/vir/to_viper.rs:
```diff
@@ -113,7 +113,7 @@ impl<'v> ToViper<'v, viper::Stmt<'v>> for Stmt {
             }
             Stmt::Exhale(ref expr, ref pos) => {
-                assert!(!pos.is_default());
+                // assert!(!pos.is_default(), "stmt with default pos: {}", self);
```
Do you have statements introduced via `vir!` that don't get their position reassigned to something sensible?

dario23 replied:
yes, it seems. is there a reason `vir::Stmt::set_pos` only actually does anything for `Stmt::PackageMagicWand`?

dario23 replied:
added `set_default_pos` calls for the statements encoding the expiration of the reference into the array as well, seems to fix it, together with adding `Stmt::Exhale` to the match in `pos`, `set_pos` in `prusti-common/src/vir/ast/stmt.rs`.

Aurel300 replied:
Yeah, that looks a bit strange. I hope that this can only manifest in UI-related tests, if at all.

Aurel300 commented on prusti-tests/tests/verify/pass/arrays/array-mut-loop.rs:
```diff
@@ -0,0 +1,19 @@
+
+    let mut i = 0;
+    let mut cont = true;
+    while cont {
```
Why this strange `cont` boolean instead of `i < 3`?

dario23 replied:
somehow didn't think that's supported. updated. i can't find the documentation that says what loop conditions are supported/unsupported right now. :-/

Aurel300 replied:
I think at this point we support basically arbitrary code in the loop guard? We don't support e.g. `for` loops, but IIUC it's because of the iterator there.

Aurel300 commented on prusti-tests/tests/verify/pass/arrays/array-mut.rs:
```diff
@@ -0,0 +1,5 @@
+fn main() {
+    let mut a = [1, 2, 3];
+    a[1] = 23;
+    assert!(a[1] == 23);
```
```suggestion
    assert!(a[0] == 1);
    assert!(a[1] == 23);
    assert!(a[2] == 3);
```

Aurel300 commented on prusti-tests/tests/verify/pass/arrays/array-ref-mut.rs:
```diff
@@ -0,0 +1,13 @@
+#![feature(nll)]
+use prusti_contracts::*;
+
```
```suggestion
```

Aurel300 commented on prusti-viper/src/encoder/mir_encoder/place_encoding.rs:
```diff
@@ -53,6 +53,26 @@ impl<'tcx> PlaceEncoding<'tcx> {
     }
 
+    /// Ok(e) => the same result try_into_expr would give
+    /// Err(e) => the array base expression
```
Please don't use the error variant for things that are not actually errors. Define a dedicated enum, or see if you can make the uses of this function just do `try_into_array_base`, returning an `Option`, otherwise falling back to `try_into_expr`.

dario23 replied:
tried to keep the duplication with `try_into_expr` in bounds, added a separate `ExprOrArrayBase` enum.

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -99,6 +124,8 @@ pub struct ProcedureEncoder<'p, 'v: 'p, 'tcx: 'v> {
     cfg_block_has_been_executed: HashMap<mir::BasicBlock, vir::LocalVar>,
     /// Magic wand generated by a call at a given location with a label used in post.
     magic_wand_at_location: HashMap<mir::Location, (String, vir::Expr, vir::Expr)>,
+    /// magic wands from array accesses
```
Please explain the 4-tuple elements.

dario23 committed:
Revert to manual fmt::Debug impl for Node

This one has less info, but the derived one is too noisy for everything
else than debugging.

dario23 committed:
Properly set position for exhale statements, re-enable assert

dario23 committed:
Set pos for array ref expiration stmts as well

dario23 committed:
Avoid separate loop variable in tests

dario23 committed:
Introduce ExprOrArrayBase enum for PlaceEncoding

dario23 committed:
Document and simplify ProcedureEncoder.array_magic_wand_at

dario23 committed:
Update error message for non-supported array RHS cases

dario23 committed:
Add a testcase for the complex array RHS error

dario23 committed:
Match Slices in PlaceEncoding

Aurel300 commented on prusti-common/src/vir/ast/expr.rs:
```diff
@@ -1198,7 +1198,7 @@ impl Expr {
         let typaram_substs = match (&target, &replacement) {
             (Expr::Local(tv, _), Expr::Local(rv, _)) => {
-                if tv.typ.is_ref() && rv.typ.is_ref() {
+                if tv.typ.is_ref() && rv.typ.is_ref() && !replacement.local_type().starts_with("Array$") {
```
Generics are messy… What exactly goes wrong if you don't have this extra condition? Does it complain somewhere later that a type is not a place?

dario23 replied:
not sure anymore. removed it, array testcases still pass. i'll do a full run of the testsuite and see, maybe it's not needed (anymore?) after all.

dario23 replied:
tests look good :-)

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -1130,160 +1138,262 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         stmts.extend(encoding_stmts);
-        Ok(stmts
+        Ok(self.set_stmts_default_pos(stmts, stmt.source_info.span))
+
```
```suggestion
```

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -1130,160 +1138,262 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+    }
+
+    fn set_stmts_default_pos(&self, stmts: Vec<vir::Stmt>, default_pos_span: Span) -> Vec<vir::Stmt> {
+        let expr_pos = self
+            .encoder
+            .error_manager()
+            .register(default_pos_span, ErrorCtxt::GenericExpression);
+        let stmt_pos = self
+            .encoder
+            .error_manager()
+            .register(default_pos_span, ErrorCtxt::GenericStatement);
```
Does this not register a lot of errors in the error manager that cannot ever be errors?

dario23 replied:
not sure, i mostly just moved the code into the separate function.

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -1130,160 +1138,262 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+
+                // expiring base is something like ref$i32, so we need .val_ref.val_int
+                let deref = self.encoder.encode_value_expr(expiring_base.clone(), expiring_ty);
+                let target_ty = if let ty::TyKind::Ref(_, target_ty, _) = expiring_ty.kind() { target_ty } else { unreachable!() };
```
This doesn't panic but maybe nicer:
```suggestion
                let target_ty = expiring_ty.peel_refs();
```

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -1130,160 +1138,262 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                    vir::Stmt::Label(new_lhs_label.clone())
+                );
+
+                let wand_rhs_patched_lhs = wand_rhs.fold_expr(
```
Isn't there already a utility function to replace a particular old label somewhere in `vir`? I remember using a similar method in the (still unmerged) closures PR, but I'm not sure if I added it.

dario23 replied:
indeed, found `Expr::map_old_expr_label`, thanks :-)

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -1860,7 +1972,8 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
             } => {
                 // will panic if attempting to encode unsupported type
-                let (encoded_lhs, pre_stmts, _, _) = self.encode_place(lhs).unwrap();
+                // TODO: ignore stmts here as well?
```
Why? Should `pre_stmts` be empty? Can this be asserted?

dario23 replied:
don't think they necessarily are; i think this comment is from when i initially introduced this and wasn't sure if we needed them blindly passed on everywhere. i'll remove the comment.

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4003,26 +4119,47 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
-                let (encoded_place, pre_stmts, ty, _) = self.encode_place(&mir_place).unwrap();
-                assert!(pre_stmts.is_empty(), "Unexpected encoding pre-statements: {:?}", pre_stmts);
-                debug!("kind={:?} mir_place={:?} ty={:?}", kind, mir_place, ty);
+                let (encoded_place, ty, _) = self.mir_encoder.encode_place(&mir_place).unwrap();
```
Add a comment why this uses `mir_encoder` instead of `self`?

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4003,26 +4119,47 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
-                        permissions.push(perm);
+                        if is_array_access {
+                            let entry = array_pred_perms.entry(encoded_place).or_insert(vir::PermAmount::Write);
+                            *entry= vir::PermAmount::Write;
```
```suggestion
                            array_pred_perms[encoded_place] = vir::PermAmount::Write;
```

dario23 replied:
they [removed the `IndexMut` impl for `HashSet`](https://stackoverflow.com/a/30414450), so it seems it's not that simple anymore, but `array_pred_perms.insert(encoded_place, vir::PermAmount::Write)` should work.

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -4463,6 +4624,86 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        let zero_le_i = vir!{ [vir::Expr::from(0)] <= [ i_var ] };
+        let i_lt_len = vir!{ [ i_var ] < [ vir::Expr::from(at.array_len) ] };
+        let i_ne_idx = vir!{ [ i_var ] != [ old(idx_val_int.clone()) ] };
+        let idx_conditions = vir!{ [zero_le_i] && ([i_lt_len] && [i_ne_idx]) };
```
(Not super important, but does the `vir!` macro require this kind of parenthesisation?)

dario23 replied:
yes it does, sadly, as `&&` is a binary thing (i.e. no repeat in and of itself), and the lhs and rhs are `tt` matchers, so we'd match `[zero_le_i] && [i_lt_len]` and then run into "unexpected token" errors.

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5426,50 +5676,150 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Mutable(None, location))?;
+
+        let usize_ty = self.encoder.env().tcx().mk_ty(ty::TyKind::Uint(ty::UintTy::Usize));
+        let idx_val_int = index.field(self.encoder.encode_value_field(usize_ty));
```
```suggestion
        let idx_val_int = vir::Expr::snap_app(index);
```

dario23 replied:
i run into [this assertion](https://github.com/viperproject/prusti-dev/blob/master/prusti-viper/src/encoder/foldunfold/requirements.rs#L355) if i try that.. :-/

Aurel300 replied:
Ok, then this either needs to be patched manually:
```suggestion
        let idx_val_int = self.encoder.patch_snapshots(vir::Expr::snap_app(index))?;
```
Or just left as is.

dario23 replied:
hm, manually patching sounds fair enough, i'll go with that.

dario23 committed:
Simplify hashmap insert

dario23 committed:
Add comment why we use mir_encoder's encode_place

dario23 committed:
Remove resolved TODO

dario23 committed:
Use Expr::map_old_expr_label

dario23 committed:
misc fixes from review

dario23 committed:
Use (manual) snapshotting for idx.val_int encoding
